cli/azd/cmd/cobra_builder.go

```diff
@@ -0,0 +1,221 @@
+package cmd
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/azure/azure-dev/cli/azd/cmd/actions"
+	"github.com/azure/azure-dev/cli/azd/cmd/middleware"
+	"github.com/azure/azure-dev/cli/azd/pkg/ioc"
+	"github.com/azure/azure-dev/cli/azd/pkg/output"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools"
+	"github.com/spf13/cobra"
+)
+
+// CobraBuilder manages the construction of the cobra command tree from nested ActionDescriptors
+type CobraBuilder struct {
+	container *ioc.NestedContainer
+	runner    *middleware.MiddlewareRunner
+}
+
+// Creates a new instance of the Cobra builder
+func NewCobraBuilder(container *ioc.NestedContainer) *CobraBuilder {
+	return &CobraBuilder{
+		container: container,
+		runner:    middleware.NewMiddlewareRunner(container),
+	}
+}
+
+// Builds a cobra Command for the specified action descriptor
+func (cb *CobraBuilder) BuildCommand(descriptor *actions.ActionDescriptor) (*cobra.Command, error) {
+	cmd := descriptor.Options.Command
+	if cmd.Use == "" {
+		cmd.Use = descriptor.Name
+	}
+
+	// Build the full command tree
+	for _, childDescriptor := range descriptor.Children() {
+		childCmd, err := cb.BuildCommand(childDescriptor)
+		if err != nil {
+			return nil, err
+		}
+
+		cmd.AddCommand(childCmd)
+	}
+
+	// Bind root command after command tree has been established
+	// This ensures the command path is ready and consistent across all nested commands
+	if descriptor.Parent() == nil {
+		if err := cb.bindCommand(cmd, descriptor); err != nil {
+			return nil, err
+		}
+	}
+
+	cb.configureActionResolver(cmd, descriptor)
+
+	return cmd, nil
+}
+
+// Configures the cobra command 'RunE' function to running the composed middleware and action for the
+// current action descriptor
+func (cb *CobraBuilder) configureActionResolver(cmd *cobra.Command, descriptor *actions.ActionDescriptor) {
+	// Only bind command to action if an action resolver had been defined
+	// and when a RunE hasn't already been set
+	if descriptor.Options.ActionResolver == nil || cmd.RunE != nil {
+		return
+	}
+
+	cmd.RunE = func(cmd *cobra.Command, args []string) error {
+		ctx := cmd.Context()
+		ctx = tools.WithInstalledCheckCache(ctx)
+
+		// Registers the following to enable injection into actions that require them
+		ioc.RegisterInstance(ioc.Global, cb.runner)
+		ioc.RegisterInstance(ioc.Global, ctx)
+		ioc.RegisterInstance(ioc.Global, cmd)
+		ioc.RegisterInstance(ioc.Global, args)
+
+		if err := cb.registerMiddleware(descriptor); err != nil {
+			return err
+		}
+
+		actionName := createActionName(cmd)
+		var action actions.Action
+		if err := cb.container.ResolveNamed(actionName, &action); err != nil {
+			return fmt.Errorf(
+				//nolint:lll
+				"failed resolving action '%s'. Ensure the ActionResolver is a valid go function that returns an `actions.Action` interface, %w",
+				actionName,
+				err,
+			)
+		}
+
+		runOptions := &middleware.Options{
+			Name:    cmd.CommandPath(),
```

Comment:
Do we want `middleware.Options.Name` to be `CommandPath` (`azd env list`) or `Use` (`list`)?

Reply:
@ellismg is correct - full command path is expected here and this is the value that the Telemetry middleware will used to hook up to the existing command event name pattern.

Reply:
@wbreza Should we rename it to `CommandPath` so that it's clearer?

Reply:
I'm fine with that.